File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def issimilar(w1,w2):
    length = len(w1)
    cnt = 0
    for i in range(length):
        if w1[i] == w2[i]:
            cnt +=1
    if cnt == length-1:
        return True
    return False
    
    
def solution(begin, target, words):
    word_length = len(begin)
    graph = [[] for _ in range(len(words))]
    words_order = {}
    # 그래프 생성
    for i in range(len(words)):
        words_order[words[i]] = i
        for j in range(i+1,len(words)):
            if issimilar(words[i],words[j]):
                graph[i].append(j)
                graph[j].append(i)
                
    print(words_order)
    # BFS
#     ### visited = [False for _ in range(word_length)]
    step = [0 for _ in range(len(words))]
    
    if target not in words_order:
        print("no key")
        return 0
    
    target_idx = words_order[target]
    
    queue = deque()
    for i in range(len(words)):
        if issimilar(words[i],begin):
            queue.append(i)
            step[i] = 1
            
    while queue:
        node = queue.popleft()
        for next_node in graph[node]:
            if step[next_node] == 0:
                step[next_node] = step[node] +1
                queue.append(next_node)
    answer = step[target_idx]
    # print(step)
    # if answer == -1:
    #     answer = 0
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
// System.out.println()
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        List<String> myList = new ArrayList(Arrays.asList(words));
        myList.add(0,begin);
        int target_idx = -1;
        // start index가 begin인 graph
        int graphLen = words.length + 1;
        int[][] graph = new int[graphLen][graphLen];
        for(int i=0 ; i < graphLen;i++){
            if (myList.get(i).equals(target)){
                target_idx = i;
            }
            for(int j=0; j < graphLen; j++){
                graph[i][j] = getDistance(myList.get(i),myList.get(j));
                // System.out.print(graph[i][j]);
            }
            // System.out.println(myList.get(i)+"\n");
        }
        
        if (!myList.contains(target)){
            return 0;
        }
        answer = bfs(target_idx,graph);
        return answer;
    }
    private int getDistance(String word1, String word2){
        int distance = word1.length();
        for(int i =0; i< word1.length(); i++){
            if (word1.charAt(i) == word2.charAt(i)){
                distance -=1;
            }
        }
        return distance;
    }
    
    private int bfs(int target_idx, int[][] graph){
        Deque<Integer> queue = new LinkedList();
        queue.addFirst(0);
        int nowNode;
        int loopCount = graph[0].length;
        int[] distance = new int[loopCount]; 
        for (int i=0;i<loopCount; i++){
            distance[i] = 1000000;
        }
        
        distance[0] = 0;
        
        while (!queue.isEmpty()){
            nowNode = queue.remove();
            for(int j =0; j< loopCount; j++){
                if(graph[nowNode][j] == 1){
                    if (distance[nowNode]+1 < distance[j]){
                        distance[j] = distance[nowNode]+1;
                        queue.addFirst(j);
                    }
                }
            }
        }
        // for (int i=0; i<loopCount;i++){
        //     System.out.print(distance[i]);
        // }
        return distance[target_idx];
    }
}